sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ObjectsShouldNotBeDisposedMoreThanOnce.cs

```diff
@@ -0,0 +1,208 @@
+            {
+                var newProgramState = programState;
+
+                var arguments = usingExpression.DescendantNodes()
+                    .OfType<ObjectCreationExpressionSyntax>()
+                    .Where(this.IsStreamDisposingType)
+                    .Select(FirstArgumentOrDefault)
+                    .WhereNotNull();
+
+                foreach (var argument in arguments)
+                {
+                    var streamSymbol = semanticModel.GetSymbolInfo(argument.Expression).Symbol;
+                    newProgramState = ProcessDisposableSymbol(newProgramState, argument.Expression, streamSymbol);
+                }
+
+                return newProgramState;
+            }
+
+            private static ArgumentSyntax FirstArgumentOrDefault(ObjectCreationExpressionSyntax objectCreation) =>
+                objectCreation.ArgumentList?.Arguments.FirstOrDefault();
+
+            private bool IsStreamDisposingType(ObjectCreationExpressionSyntax objectCreation) =>
+                semanticModel.GetSymbolInfo(objectCreation.Type)
+                    .Symbol
+                    .GetSymbolType()
+                    .DerivesOrImplementsAny(typesDisposingUnderlyingStream);
+
+            private ProgramState VisitInvocationExpression(InvocationExpressionSyntax instruction, ProgramState programState)
+            {
+                var newProgramState = programState;
+
+                var disposeMethodSymbol = semanticModel.GetSymbolInfo(instruction).Symbol as IMethodSymbol;
+                if (disposeMethodSymbol.IsIDisposableDispose())
+                {
+                    var disposable = ((MemberAccessExpressionSyntax)instruction.Expression).Expression;
+
+                    var disposableSymbol = semanticModel.GetSymbolInfo(disposable).Symbol;
+
+                    newProgramState = ProcessDisposableSymbol(newProgramState, disposable, disposableSymbol);
+                }
+
+                return newProgramState;
+            }
+
+            private ProgramState ProcessDisposableSymbol(ProgramState programState, SyntaxNode instruction, ISymbol disposableSymbol)
+            {
+                if (disposableSymbol == null) // DisposableSymbol is null when we invoke an array element
+                {
+                    return programState;
+                }
+
+                if (disposableSymbol.HasConstraint(DisposableConstraint.Disposed, programState))
+                {
+                    ObjectDisposed?.Invoke(this, new ObjectDisposedEventArgs(disposableSymbol.Name, instruction.GetLocation()));
+                    return programState;
+                }
+
+                // We should not replace Null constraint because having Disposed constraint
+                // implies having NotNull constraint, which is incorrect.
+                if (!disposableSymbol.HasConstraint(ObjectConstraint.Null, programState))

```

Comment:
I would have negated this condition (especially with the comment which made me expect the opposite condition).